osu.Server.Spectator/Hubs/ScoreUploader.cs

```diff
@@ -33,22 +28,29 @@ public double UploadInterval
         private readonly ConcurrentQueue<UploadItem> queue = new ConcurrentQueue<UploadItem>();
         private readonly IDatabaseFactory databaseFactory;
         private readonly IScoreStorage scoreStorage;
-        private readonly Timer timer;
-        private readonly CancellationTokenSource timerCancellationSource;
-        private readonly CancellationToken timerCancellationToken;
+        private readonly CancellationTokenSource cancellationSource;
+        private readonly CancellationToken cancellationToken;
 
         public ScoreUploader(IDatabaseFactory databaseFactory, IScoreStorage scoreStorage)
         {
             this.databaseFactory = databaseFactory;
             this.scoreStorage = scoreStorage;
 
-            timerCancellationSource = new CancellationTokenSource();
-            timerCancellationToken = timerCancellationSource.Token;
+            cancellationSource = new CancellationTokenSource();
+            cancellationToken = cancellationSource.Token;
 
-            timer = new Timer(50);
-            timer.AutoReset = false;
-            timer.Elapsed += (_, _) => Task.Run(Flush);
-            timer.Start();
+            Task.Run(runFlushLoop);
+        }
+
+        private void runFlushLoop()
+        {
+            while (!queue.IsEmpty || !cancellationToken.IsCancellationRequested)
```

Comment:
Doesn't checking for empty here mean that if the queue ever becomes empty then the uploader will stop flushing forevermore?

Reply:
I'm open to rewriting the conditional to be less confusing, but the logic currently there does check out as far as i can tell.

Reply:
Oh wait, it's an `||` not an `&&`. Misread.

Misled me because the `queue.IsEmpty` is basically a dead condition _until_ cancellation happens.